src/v/storage/segment_appender.cc

```diff
@@ -445,66 +446,113 @@ void segment_appender::dispatch_background_head_write() {
 
     const size_t start_offset = ss::align_down<size_t>(
       _committed_offset, _head->alignment());
-    const size_t expected = _head->dma_size();
-    const char* src = _head->dma_ptr();
-    vassert(
-      expected <= _chunk_size,
-      "Writes can be at most a full segment. Expected {}, attempted write: {}",
-      _chunk_size,
-      expected);
+
     // accounting synchronously
+    const auto prior_co = _committed_offset;
     _committed_offset += _head->bytes_pending();
     _bytes_flush_pending -= _head->bytes_pending();
+
+    inflight_write entry{
+      .chunk = _head,
+      .chunk_begin = _head->pending_aligned_begin(),
+      .chunk_end = _head->pending_aligned_end(),
+      .file_start_offset = start_offset,
+      .committed_offset = _committed_offset};
+
     // background write
     _head->flush();
+
+    entry.full = _head->is_full();
+
+    // defer to account for early exits
+    auto reap_head = ss::defer([this, full = entry.full] {
+        if (full) {
+            /*
+             * If _head is full then this is the last write to this chunk, so we
+             * clear out the head pointer synchronously here, then release it
+             * back into the chunk cache after the write completes. Otherwise,
+             * leave it in place so that new appends may accumulate. this
+             * optimization is meant to avoid redhydrating the chunk on append
+             * following a flush when the head has pending bytes and a write is
+             * dispatched.
+             */
+            _head = nullptr;
+            /*
+             * When the head becomes full it still needs to be properly
+             * sequenced with earlier writes to the same head, but no future
+             * writes to same head head are possible so the dependency chain is
+             * reset for the next head.
+             */
+            _prev_head_write = ss::make_lw_shared<ssx::semaphore>(
+              1, head_sem_name);
+        }
+    });
+
+    if (!_inflight.empty() && _inflight.back()->try_merge(entry, prior_co)) {
```

Comment:
I don't know if you are doing to like this reason but I did not want to introduce another level of indentation for the rest of the function, in part because I don't like deep indentation and in part because it tends to make reviewing harder due to every line inside being different (may be solved by ignoring whitespace) and clang-format making different line-breaking decisions.

That said, the formatting ended up changing a lot so may I'll just apply De Morgan after all.

Reply:
> I don't know if you are doing to like this reason but I did not want to introduce another level of indentation

Haha, that part of clang-format is a bit annoying indeed.

Reply:
@StephanDollberg - I ended up avoiding the defer but still being able to use the early-out by re-arranging things a bit.